### PR TITLE
Rewrite pool interface without actors

### PR DESCRIPTION
**Pull request by jrudolph**
* adding `PoolInterface` as yet another abstraction layer to get the sprawling layers of the client implementation under control.
 * reimplement `PoolInterface` as a `GraphStage` instead of using `ActorPublisher`/`ActorSubscriber`

Refs #2617

### CODE REVIEW

**Comment by jrudolph** [comment]
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+private[http] object PoolInterface {
+  private[http] sealed trait ShutdownReason
+  private[http] object ShutdownReason {
+    case object ShutdownRequested extends ShutdownReason
+    case object IdleTimeout extends ShutdownReason
+  }
+
+  def create(gateway: PoolGateway, parent: ActorRefFactory)(implicit fm: Materializer): PoolInterface = {
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala]:
```suggestion
  def apply(gateway: PoolGateway, parent: ActorRefFactory)(implicit fm: Materializer): PoolInterface = {
```
?

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 2]
```diff
@@ -260,16 +260,20 @@ private[http] object StreamUtils {
...
-  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Unit], Future[Unit])] {
-    def strictM: (Future[Unit], Future[Unit]) = (Future.successful(()), Future.successful(()))
-    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Unit], Future[Unit])) = {
+  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])] {
+    def strictM: (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]]) = (Future.successful(()), Future.successful(()), Future.successful(None))
+    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])) = {
       val materializationPromise = Promise[Unit]()
+      val killSwitchPromise = Promise[Option[UniqueKillSwitch]]()
+      val newSource0 =
+        source.via(Flow.fromGraph(KillSwitches.single[T]).mapMaterializedValue(ks => killSwitchPromise.trySuccess(Some(ks))))
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: Do we expect adding the killswitch here to affect performance?

Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: Probably not more than the already termination capturing, but maybe I could fold it into the existing stage. In any case, cost is amortized to some degree at the OS receive buffers. We always read up to the configured direct buffer size from the OS so if you read less you'll get more data in the next run.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala [hunk 3]
```diff
@@ -187,7 +187,7 @@ class HostConnectionPoolSpec extends AkkaSpec(
 
         val streamResult = chunks.runWith(Sink.ignore)
         Await.ready(streamResult, 3.seconds)
-        streamResult.value.get.failed.get.getMessage shouldEqual "Substream Source cannot be materialized more than once"
+        streamResult.value.get.failed.get.getMessage shouldEqual "Connection was closed while response was still in-flight"
```
Comment (raboof) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/HostConnectionPoolSpec.scala]: Nice

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala [hunk 4]
```diff
@@ -302,6 +302,41 @@ abstract class ConnectionPoolSpec(poolImplementation: PoolImplementation) extend
...
+      val (Success(r1), 42) = responseOut.expectNext()
+      val responseEntityProbe = ByteStringSinkProbe()
+      r1.entity.dataBytes.runWith(responseEntityProbe.sink)
+      responseEntityPub.sendNext(ByteString("Hello"))
+      responseEntityProbe.expectUtf8EncodedString("Hello")
+
+      Await.result(gateway.poolStatus(), 1500.millis.dilated).get shouldBe a[PoolInterfaceRunning]
+      Thread.sleep(500)
+
+      // afterwards it should still be running because a response is still being delivered
```
Comment (raboof) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala]: Hmm, so a slow reader would hold up closing the pool. I guess that's OK, and this is probably guarded by another timeout somewhere.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala [hunk 5]
```diff
@@ -549,7 +558,7 @@ private[client] object NewHostConnectionPool {
 
         override def onUpstreamFinish(): Unit = {
           log.debug("Pool upstream was completed")
-          super.onDownstreamFinish()
+          super.onUpstreamFinish()
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala]: :hushed: :+1:

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala [hunk 6]
```diff
@@ -47,14 +51,16 @@ private[http] final class PoolMasterActor extends Actor with ActorLogging {
    * @param fm the materializer to use for this pool
    * @return the newly created actor ref
    */
-  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): PoolInterface = {
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala]: Maybe rename

**Comment by raboof** [comment]
Pull request against this branch at https://github.com/jrudolph/akka-http/pull/1

**Comment by raboof** [comment]
Current status:
* Please review jrudolph#1
* https://github.com/akka/akka-http/pull/2630#discussion_r313260437
* rebase for conflict

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala (hunk 6 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala]: Done.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala (hunk 4 above)
Comment (jrudolph) [reply akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala]: Yes, exactly, we fix #1847 with that new functionality.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolInterface.scala]: Done.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala [hunk 7]
```diff
@@ -47,14 +51,16 @@ private[http] final class PoolMasterActor extends Actor with ActorLogging {
...
-    val ref = context.actorOf(PoolInterfaceActor.props(gateway), PoolInterfaceActor.name.next())
-    poolStatus += gateway -> PoolInterfaceRunning(ref)
-    poolInterfaces += ref -> gateway
-    context.watch(ref)
+    val interface = PoolInterface.create(gateway, context)
+    poolStatus += gateway -> PoolInterfaceRunning(interface)
+    poolInterfaces += interface -> gateway
+    interface.whenShutdown.onComplete { reason => self ! HasBeenShutdown(interface, reason) }(context.dispatcher)
+    interface
+    // context.watch(ref) - how to replace that one?
```
[the remark this answers is not in the document]
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala]: Indeed, left over from an earlier version.

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala (hunk 2 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: I'll do this improvement in a follow-up PR.

**Comment by jrudolph** [comment]
PLS BUILD

**Pull request merged**